lib/improver/spotdata/extrema.py

```diff
@@ -0,0 +1,255 @@
+
+    e.g.
+
+    UTC Coord  :  12   13   14   15   16
+    Data: Site 1  300  302  296  294  290 (UTC offset = -2)
+          Site 2  280  282  283  280  279 (UTC offset = +2)
+
+    Data redistributed according to UTC offset to sit on local time.
+
+    Local times:  10   11   12   13   14   15   16   17   18
+    Data: Site 1  300  302  296  294  290  -    -    -    -
+          Site 2  -    -    -    -    280  282  283  280  279
+
+    There will be missing but masked data in locations which are not forecast
+    at the given local time, e.g. a SpotData site at a UTC location for a 12Z
+    run will have no forecast temperatures on the local time axis for times
+    earlier than 12Z.
+
+    Maxima/Minima can then be calculated on local time, as makes sense for
+    quantities such as maximum in day.
+
+    New index is obtained as:
+        Site_UTC_offset + 12 + UTC_coordinate_index
+
+        +12 ensures that UTC offset -12 sits at an index of 0 etc.
+
+        UTC_coordinate_index is the data's index on the original coordinate
+        axis. So imagining a model run with a forecast reference time of 12Z,
+        a temperature at 14Z UTC will sit at index=2. At a SpotData site with
+        a UTC offset of -6, the index will become 8:
+
+   Index: 0  1  2  3  4  5  6  7  8  9  10 11 12
+   UTC    12 13 14 15 16 17 18 19 20 21 22 23 00
+   Data :       X
+   Local: 00 01 02 03 04 05 06 07 08 09 10 11 12
+   Data :                         X
+
+
+    Args:
+    -----
+    cube : iris.cube.Cube
+        A cube of site data resulting from the SpotData extraction process.
+
+    Returns:
+    --------
+    cube : iris.cube.Cube
+        A cube with localised time coordinates.
+
+    """
+    # Ensure time coordinate is in hours.
+    hour_coordinates = cube.coord('time')
+    hour_coordinates.convert_units('hours since 1970-01-01 00:00:00')
+
+    # Calculate local time range which spans earliest time UTC -12 for far west
+    # to latest time UTC +14 in far east. Ignores half hour time zones.
+    local_time_min = int(hour_coordinates.points[0]-12)
+    local_time_max = int(hour_coordinates.points[-1]+14)
+    local_times = range(local_time_min, local_time_max+1, 1)
+
+    # Create iris.coord.DimCoord of local times.
+    local_time_coord = DimCoord(local_times, standard_name='time',
+                                units=hour_coordinates.units)
+
+    # Create empty array to contain extrema data.
+    new_data = np.full((len(local_times), cube.data.shape[1]), np.nan)
+
+    # Create ascending indices to help with filling new_data array.
+    n_sites = cube.data.shape[1]
+    row_index = range(0, n_sites)
+
+    # Loop through times in UTC and displace data in array so that each datum
+    # sits at its local time.
+    for i_time in range(len(hour_coordinates.points)):
+        indices = cube.coord('utc_offset').points.astype(int) + 12 + i_time
+        new_data[indices, row_index] = cube.data[i_time]
```

Comment:
After testing this, I can see that this function is currently only usable by an input cube of quite a specific format where `time` and `index` are the dimension coordinates, as this function is quite prescriptive about the shape of the data that is being worked with i.e. a 2d data array. I assume that as far as currently imagined, `spotdata` cubes will only have these dimensions and other dimensions e.g. `realization`, `height` won't exist at this point?

For reference, I tested this by modifying the `Test_make_local_time_cube` test class as follows:
```python
class Test_make_local_time_cube(Test_extrema):
    """Test time localisation function."""

    def test_time_coord(self):
        """
        Test that a new time_coord has been constructed that spans all local
        times. UTC-12 to UTC+14 is the maximum range of timezones.

        """
        from cf_units import Unit
        realizations=np.array([0, 1, 2])
        timesteps=1
        y_dimension_length=3
        x_dimension_length=3
        data = (np.tile(np.linspace(-45.0, 45.0, 9), 3).reshape(3, 1, 3, 3) +
                273.15)
        data[0] -= 2
        data[1] += 2
        data[2] += 4
        phenomenon_standard_name = "air_temperature"
        phenomenon_units = "K"
        print "data = ", data.shape
        cube = Cube(data, standard_name=phenomenon_standard_name,
                    units=phenomenon_units)
        cube.add_dim_coord(DimCoord(realizations, 'realization',
                                    units='1'), 0)
        time_origin = "hours since 1970-01-01 00:00:00"
        calendar = "gregorian"
        tunit = Unit(time_origin, calendar)
        cube.add_dim_coord(DimCoord(np.linspace(402192.5, 402292.5, timesteps),
                                    "time", units=tunit), 1)
        cube.add_dim_coord(DimCoord(np.linspace(-45.0, 45.0, y_dimension_length),
                                    'latitude', units='degrees'), 2)
        cube.add_dim_coord(DimCoord(np.linspace(120, 180, x_dimension_length),
                                    'longitude', units='degrees'), 3)
        cube.add_aux_coord(AuxCoord(np.array([0]), long_name='utc_offset',
                              units='hours'), 1)
        expected_first_time = self.time_coord.points[0] - 12
        expected_last_time = self.time_coord.points[-1] + 14
        #result = make_local_time_cube(self.cube)
        result = make_local_time_cube(cube)
        self.assertEqual(result.coord('time').points[0],
                         expected_first_time)
        self.assertEqual(result.coord('time').points[-1],
                         expected_last_time)
```